### PR TITLE
Add default certificate caching behaviour

### PR DESCRIPTION
**Pull request by piotr-gregor**
By setting

stir_shaken_vs_set_default_cert_caching(ss, vs, 1);
stir_shaken_vs_set_cache_expire_seconds(...);
stir_shaken_vs_set_cache_dir(...);

certificate caching inside given dir is turned on. This will use default implementation of shaken callback: stir_shaken_vs_default_callback().

There are three independent setings now that cross some logic inside the core of the lib and must be tested with each other: 
- stir_shaken_vs_set_callback
- stir_shaken_vs_set_x509_cert_path_check
- stir_shaken_vs_set_default_cert_caching

This results in 8 additional VS setups that must be tested + those with caching on must be tested with and without cert available in cache, so it's 2 * 2 * 3 = 12 in total.

### CODE REVIEW

File: src/stir_shaken_ssl.c [hunk 1]
```diff
@@ -2479,6 +2481,7 @@ stir_shaken_status_t stir_shaken_do_sign_data_with_digest(stir_shaken_context_t
         EC_KEY			*ec_key = NULL;
         const BIGNUM	*ec_sig_r = NULL;
         const BIGNUM	*ec_sig_s = NULL;
+		const unsigned char **tmpsig_p = (const unsigned char **) &tmpsig;
```
Comment (crienzo) [inline src/stir_shaken_ssl.c:2484]: inconsistent indentation

File: src/stir_shaken_verify.c [hunk 2]
```diff
@@ -265,6 +279,8 @@ stir_shaken_status_t stir_shaken_jwt_fetch_or_download_cert(stir_shaken_context_
 
 	stir_shaken_destroy_http_request(&http_req);
 
+anyway:
```
Comment (crienzo) [inline src/stir_shaken_verify.c:282]: This label looks to be unused

File: src/stir_shaken_service.c [hunk 3]
```diff
@@ -867,8 +867,32 @@ stir_shaken_status_t stir_shaken_vs_set_callback(struct stir_shaken_context_s *s
 		return STIR_SHAKEN_STATUS_TERM;
 	}
 
+	if (vs->settings.default_cert_caching) {
+		stir_shaken_set_error(ss, "Callback set callback when default certificate cahing behaviuor is turned on (use wither callback or default cert caching)", STIR_SHAKEN_ERROR_VS_INVALID_SETTINGS);
```
Comment (crienzo) [inline src/stir_shaken_service.c:871]: caching behaviour

File: src/stir_shaken_service.c [hunk 4]
```diff
@@ -867,8 +867,32 @@ stir_shaken_status_t stir_shaken_vs_set_callback(struct stir_shaken_context_s *s
...
+
+stir_shaken_status_t stir_shaken_vs_set_callback_user_data(struct stir_shaken_context_s *ss, stir_shaken_vs_t *vs, void *user_data)
+{
+	if (!vs) {
+		stir_shaken_set_error(ss, "Verification service missing", STIR_SHAKEN_ERROR_VS_MISSING_9);
+		return STIR_SHAKEN_STATUS_TERM;
+	}
+
+	if (vs->settings.default_cert_caching) {
+		stir_shaken_set_error(ss, "Callback set callback when default certificate cahing behaviuor is turned on (use wither callback or default cert caching)", STIR_SHAKEN_ERROR_VS_INVALID_SETTINGS);
```
Comment (crienzo) [inline src/stir_shaken_service.c:890]: whether

File: src/stir_shaken_service.c [hunk 5]
```diff
@@ -943,6 +1027,280 @@ stir_shaken_status_t stir_shaken_vs_sih_verify(stir_shaken_context_t *ss, stir_s
...
+	} else {
+		ss->callback = vs->callback;
+		ss->user_data = vs->user_data;
+	}
+
 	return stir_shaken_sih_verify_ex(ss, sih, cert_out, passport_out, vs->store, vs->settings.x509_cert_path_check, vs->settings.connect_timeout_s);
 }
+
+static unsigned long hash_to_long(const char *str)
+{
```
Comment (crienzo) [inline src/stir_shaken_service.c:1042]: What is the risk of collisions with this hashing function?

File: src/stir_shaken_service.c [hunk 6]
```diff
@@ -943,6 +1027,280 @@ stir_shaken_status_t stir_shaken_vs_sih_verify(stir_shaken_context_t *ss, stir_s
...
+
+	if (!vs) {
+		stir_shaken_set_error(ss, "Verification service missing", STIR_SHAKEN_ERROR_VS_MISSING_13);
+		fprintif(STIR_SHAKEN_LOGLEVEL_HIGH, "(VS default fetching) Error. VS missing\n");
+		return STIR_SHAKEN_STATUS_ACTION_ERROR;
+	}
+
+	arg = &ss->callback_arg;
+
+	if (-1 == get_cert_name_hashed(ss, arg->cert.public_url, vs->settings.cache_dir, cert_full_path, STIR_SHAKEN_BUFLEN)) {
```
Comment (crienzo) [inline src/stir_shaken_service.c:1095]: due to possibility of collision- do you verify the cached cert is what is expected?

**Review by crienzo — CHANGES_REQUESTED** [review]

File: src/stir_shaken_service.c (hunk 4 above)
Comment (piotr-gregor) [reply src/stir_shaken_service.c:890]: either

File: src/stir_shaken_service.c (hunk 5 above)
Comment (piotr-gregor) [reply src/stir_shaken_service.c:1042]: Assuming urls are hashed it's good enough.

File: src/stir_shaken_service.c (hunk 6 above)
Comment (piotr-gregor) [reply src/stir_shaken_service.c:1095]: Can't verify anything really, since all you get is a URL.

**Pull request closed without merging**